pkg/cmd/pulumi/plugin/plugin_prune_test.go

```diff
@@ -0,0 +1,457 @@
+	dryRun                 bool
+	yes                    bool
+	latestOnly             bool
+	deletePlugin           func(workspace.PluginInfo) error
+	Stdout                 io.Writer
+}
+
+func (cmd *testPluginPruneCmd) Run() error {
+	// Ensure we have a writer for stdout
+	if cmd.Stdout == nil {
+		cmd.Stdout = os.Stdout
+	}
+
+	// Get all plugins
+	plugins, err := cmd.getPluginsWithMetadata()
+	if err != nil {
+		return err
+	}
+
+	if len(plugins) == 0 {
+		cmd.diag.Infof(diag.Message("", "no plugins found to prune"))
+		return nil
+	}
+
+	// Group plugins by kind, name, and major version
+	groups := make(map[string][]workspace.PluginInfo)
+	for _, plugin := range plugins {
+		// Skip bundled plugins - we don't want to mess with these
+		if workspace.IsPluginBundled(plugin.Kind, plugin.Name) {
+			continue
+		}
+
+		// Create a key that includes kind, name, and major version (if available)
+		var key string
+		if plugin.Version != nil {
+			if cmd.latestOnly {
+				// When using latestOnly, only group by kind and name
+				key = fmt.Sprintf("%s|%s", plugin.Kind, plugin.Name)
+			} else {
+				// Group by kind, name, and major version
+				key = fmt.Sprintf("%s|%s|v%d", plugin.Kind, plugin.Name, plugin.Version.Major)
+			}
+		} else {
+			// If no version, just use kind and name
+			key = fmt.Sprintf("%s|%s", plugin.Kind, plugin.Name)
+		}
+
+		groups[key] = append(groups[key], plugin)
+	}
+
+	// For each group, identify plugins to remove (all but the latest version)
+	toRemove := make([]workspace.PluginInfo, 0, len(plugins))
+	var totalSizeRemoved uint64
+
+	for _, group := range groups {
+		if len(group) <= 1 {
+			// Only one version, keep it
+			continue
+		}
+
+		// Sort versions in descending order
+		sort.Slice(group, func(i, j int) bool {
+			// If either version is nil, keep the one with a version
+			if group[i].Version == nil {
+				return false
+			}
+			if group[j].Version == nil {
+				return true
+			}
+			// Otherwise sort by version (newer versions first)
+			return group[i].Version.GT(*group[j].Version)
+		})
+
+		// Remove the rest
+		for i := 1; i < len(group); i++ {
+			toRemove = append(toRemove, group[i])
+			totalSizeRemoved += group[i].Size
+		}
+	}
+
+	if len(toRemove) == 0 {
+		cmd.diag.Infof(diag.Message("", "no plugins found to prune"))
+		return nil
+	}
+
+	if cmd.dryRun {
+		fmt.Fprintln(cmd.Stdout, "Dry run - no changes made")
+		fmt.Fprintf(cmd.Stdout, "Would remove %d plugins, reclaiming %s\n",
+			len(toRemove), humanize.Bytes(totalSizeRemoved))
+		return nil
+	}
+
+	// Remove the plugins
+	var failed int
+	for _, plugin := range toRemove {
+		versionStr := "n/a"
+		if plugin.Version != nil {
+			versionStr = plugin.Version.String()
+		}
+
+		if err := cmd.deletePlugin(plugin); err == nil {
+			fmt.Fprintf(cmd.Stdout, "removed: %s %s v%s\n", plugin.Kind, plugin.Name, versionStr)
+		} else {
+			fmt.Fprintf(cmd.Stdout, "failed to remove: %s %s v%s: %v\n", plugin.Kind, plugin.Name, versionStr, err)
```

Comment:
We should print the error message here to give the user an indication how to fix this.  Potentially also the plugin location.